src/lib/cypress-testrail-reporter.ts

```diff
@@ -3,89 +3,155 @@ import * as moment from 'moment';
-        console.log('\n', chalk.magenta.underline.bold('(TestRail Reporter)'));
-        console.warn(
-          '\n',
-          'No testcases were matched. Ensure that your tests are declared correctly and matches Cxxx',
-          '\n'
-        );
-        this.testRail.deleteRun();
+      runner.on('pass', test => {
+        this.submitResults(Status.Passed, test, `Execution time: ${test.duration}ms`);
+      });
 
-        return;
-      }
+      runner.on('fail', (test, err) => {
+        this.submitResults(Status.Failed, test, `${err.message}`);
+      });
 
-      // publish test cases results & close the run
-      this.testRail.publishResults(this.results)
-        .then(() => this.testRail.closeRun());
-    });
+      runner.on('retry', test => {
+        this.submitResults(Status.Retest, test, 'Cypress retry logic has been triggered!');
+      });
+
+      runner.on('end', () => {
+        /**
+         * When we reach final number of spec files 
+         * we should close test run at the end
+         */
+        var numSpecFiles = this.testRailValidation.countTestSpecFiles();
+        var counter = TestRailCache.retrieve('runCounter');
+        if (numSpecFiles.length > counter) {
+          runCounter++
+        } else {
+          this.testRailApi.closeRun();
+          // Remove testrail-cache.txt file at the end of execution
+          TestRailCache.purge();
+        }
+
+        /**
+         * Notify about the results at the end of execution
+         */
+        if (this.results.length == 0) {
+          TestRailLogger.warn('No testcases were matched with TestRail. Ensure that your tests are declared correctly and titles contain matches to format of Cxxxx');
+        } else {
+          this.runId = TestRailCache.retrieve('runId');
+          var path = `runs/view/${this.runId}`;
+          TestRailLogger.log(`Results are published to ${chalk.magenta(`https://${this.reporterOptions.host}/index.php?/${path}`)}`);
+        }
+      });
+    }
   }
 
-  private validate(options, name: string) {
-    if (options == null) {
-      throw new Error('Missing reporterOptions in cypress.json');
+  /**
+   * Ensure that after each test results are reported continuously
+   * Additionally to that if test status is failed or retried there is possibility 
+   * to upload failed screenshot for easier debugging in TestRail
+   * Note: Uploading of screenshot is configurable option
+   */
+  public submitResults (status, test, comment) {
+    const caseIds = titleToCaseIds(test.title);
+    if (caseIds.length > 0) {
+      var caseResults = caseIds.map(caseId => {
+        return {
+          case_id: caseId,
+          status_id: status,
+          comment: comment,
+        };
+      });
     }
-    if (options[name] == null) {
-      throw new Error(`Missing ${name} value. Please update reporterOptions in cypress.json`);
+    if (caseResults === undefined) {
+      //Do nothing since no tests were matched with TestRail
+    } else {
```

Comment:
Since `titleToCaseIds` will always return array, you can just have a quick check here instead of doing those 2 `if` statements. Something like:
```
const caseIds = titleToCaseIds(test.title);
if (!caseIds.length) {
  return
}
const caseResults = caseIds.map(...)
 .
.
.
```